sandbox/lua/modules/cbuf_utils.lua

```diff
@@ -0,0 +1,151 @@
+            `seconds_per_row`, `columns`, and `column_info`.
+          - rows:
+            An array or nested row arrays, where each row array contains the
+            column values, in order, either as a number or as the string
+            literal "nan". Note that the last row should typically not be
+            used, as it is probably incomplete.
+        - If parsing fails, or the header data isn't valid, nil is returned.
+
+**get_start_idx(last_time, headers)**
+
+    Given the timestamp of a previous cbuf payload and the headers from a more
+    current cbuf payload, calculates the index of the first row of new data in
+    the more current cbuf data set.
+
+    *Arguments*
+        - last_time (number)
+            The `time` value retrieved from the headers table of the earlier
+            cbuf data.
+        - headers (table)
+            The headers table (as returned from `parse_cbuf`) of the current
+            cbuf data.
+
+    *Return*
+
+        - If calculations are successful, will return a positive integer
+          number value corresponding to the index of the first row that
+          contains new data since the previous cbuf time.
+        - If the new cbuf time is the same as the provided last_time, will
+          return -2.
+        - If calculations fail (because last_time is more recent than new
+          time, or because the time interval isn't exactly divisible by the
+          cbuf's seconds_per_row value), will return -1.
+--]]
+
+local cjson = require "cjson"
+local string = require "string"
+local pcall = pcall
+local tonumber = tonumber
+
+local M = {}
+setfenv(1, M) -- Remove external access to contain everything in the module.
+
+local function check_headers(headers)
+    if (not headers.time) or (not headers.rows) or (not headers.seconds_per_row)
+        or (not headers.columns) or (not headers.column_info) then
+
+        return
+    end
+    if headers.columns ~= #headers.column_info then
+        return
+    end
+    return true
+end
+
+-- [[ Public Interface --]]
+
+function parse_cbuf(cbuf_text)
+    local rows = {}
+    local headersText, headers, ok
+    local first = true
+    for row in string.gmatch(cbuf_text, "[^\n]+") do
+        if first == true then
+            -- Parse the headers.
+            headersText = row
+            ok, headers = pcall(cjson.decode, headersText)
+            if not ok then
+                return
+            end
+            -- Skip annotations line, if it exists.
+            if not headers.annotations then
+                if not check_headers(headers) then
+                    return
+                end
+                first = false
+            end
+        else
+            -- Parse the row values
+            local values = {}
+            for value in string.gmatch(row, "[^\t]+") do
+                if value ~= "nan" then
+                    value = tonumber(value)
+                    if not value then
+                        value = "nan"
```

Comment:
if someone was using this parser and round trip the data back into a cbuf it would not work.
```lua
value = "nan"
cb:set(ts, 1, value)
```